### PR TITLE
Update completers tutorial

### LINKED ISSUE #4285: add tutorial for completers with `CompletionContext` and add some standard completers
The completers tutorial [here](https://xon.sh/dev/tutorial_completers.html) need to be updated to reflect completers with `CompletionContext` and the utility methods.

Also `xonsh ` should have some standard completer classes like [`ptk`](https://python-prompt-toolkit.readthedocs.io/en/master/pages/asking_for_input.html#nested-completion) does. Especially `NestedCompleter` can be used for setting completions for most of the CLIs.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by daniel-shimon**
Promotes #4285
Promotes #4305

Also adds `$XONSH_TRACE_COMPLETIONS` to allow completions tracing

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

File: xonsh/parsers/completion_context.py [hunk 1]
```diff
@@ -27,32 +27,47 @@
 
 
 class CommandArg(NamedTuple):
+    """An argument for a command"""
+
     value: str
```
Comment (jnoortheen) [inline xonsh/parsers/completion_context.py:32]: can this be replaced with Attributes section?

**Review by jnoortheen — CHANGES_REQUESTED** [review]

File: xonsh/parsers/completion_context.py (hunk 1 above)
Comment (daniel-shimon) [reply xonsh/parsers/completion_context.py:32]: What do you mean?

Comment (jnoortheen) [reply xonsh/parsers/completion_context.py:32]: https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring
Like here mentioned Attributes section

Comment (daniel-shimon) [reply xonsh/parsers/completion_context.py:32]: Oh, no because this is how `NamedTuples` are used

Comment (jnoortheen) [reply xonsh/parsers/completion_context.py:32]: Good then

**Review by jnoortheen — APPROVED** [review]

File: xonsh/parsers/completion_context.py [hunk 2]
```diff
@@ -68,23 +83,44 @@ def raw_prefix(self):
...
 class CompletionContext(NamedTuple):
+    """
+    The object containing the current completion context.
+    """
+
     command: Optional[CommandContext] = None
+    """
+    The current command.
+    This will be ``None`` when we can't be completing a command, e.g. ``echo @(<TAB>``.
+    """
```
Comment (jnoortheen) [inline xonsh/parsers/completion_context.py:118]: I think you wanted to mention the reverse

Comment (daniel-shimon) [reply xonsh/parsers/completion_context.py:118]: Nope - in the example we aren't completing a command, for example `echo @(ls -l)` won't work

**Comment by daniel-shimon** [comment]
@jnoortheen Thanks for the review!

**Pull request merged**